File: frontend/lib/analyze/phraseRules.ts
import type { EvidenceItem, ExtractedOpportunity } from "@/types/analysis";

/**
 * Deterministic, client-side extraction engine.
 *
 * Mirrors the rule layer of backend/services/extraction/phrase_rules.py so
 * the demo works end-to-end with zero network calls. When the backend LLM
 * extraction is wired in, this becomes the offline fallback — both produce
 * the same ExtractedOpportunity shape.
 *
 * Every detection records sentence-level evidence with a confidence score.
 */

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Find the full sentence containing a match index (for evidence display). */
function sentenceAt(text: string, index: number): string {
  const boundary = /[.!?\n]/;
  let start = index;
  while (start > 0 && !boundary.test(text[start - 1])) start--;
  let end = index;
  while (end < text.length && !boundary.test(text[end])) end++;
  return text.slice(start, Math.min(end + 1, text.length)).trim();
}

type Detection = { match: string; source: string; index: number };

function detect(text: string, pattern: RegExp): Detection | null {
  const m = pattern.exec(text);
  if (!m) return null;
  return { match: m[0], source: sentenceAt(text, m.index), index: m.index };
}

// ---------------------------------------------------------------------------
// Patterns (ordered roughly by severity within each field)
// ---------------------------------------------------------------------------

const P = {
  citizensOnly:
    /\b(?:U\.?S\.?|United States)\s+citizens?(?:\s+(?:or|and)\s+(?:U\.?S\.?\s+)?(?:lawful\s+)?permanent\s+residents?)?\s+only\b|\bmust\s+be\s+(?:a\s+)?(?:U\.?S\.?|United States)\s+citizen\b|\bopen\s+only\s+to\s+(?:U\.?S\.?|United States)\s+citizens?\b|\b(?:U\.?S\.?\s+)?citizenship\s+(?:is\s+)?required\b/i,
  permanentResidentsOnly:
    /\b(?:lawful\s+)?permanent\s+residents?\s+only\b|\bmust\s+be\s+a\s+(?:lawful\s+)?permanent\s+resident\b|\bgreen\s*card\s+(?:holders?\s+only|required)\b/i,
  citizensOrPR:
    /\b(?:U\.?S\.?|United States)\s+citizens?\s+(?:or|and)\s+(?:lawful\s+)?permanent\s+residents?\b/i,
  workAuth:
    /\b(?:must\s+be\s+|applicants?\s+must\s+be\s+)?(?:eligible|authorized)\s+to\s+work\s+in\s+the\s+(?:U\.?S\.?|United States)\b|\bwork\s+authorization\s+(?:is\s+)?required\b|\bproof\s+of\s+work\s+authorization\b|\bemployment\s+eligibility\s+verification\b|\bE-?Verify\b/i,
  cptOpt: /\bCPT\b|\bOPT\b|\bcurricular\s+practical\s+training\b|\boptional\s+practical\s+training\b/i,
  ssn: /\bSSN\b|\bsocial\s+security\s+number\b/i,
  paid: /\bpaid\b|\bstipend\b|\bsalary\b|\bcompensat(?:ion|ed)\b|\bhourly\s+(?:rate|wage)\b|\$\s?\d[\d,]*/i,
  unpaid: /\bunpaid\b|\bvolunteer\b|\bno\s+compensation\b|\bnot\s+(?:a\s+)?paid\b/i,
  worldwide:
    /\bopen\s+(?:to\s+(?:students\s+)?worldwide|worldwide|globally|to\s+all\s+countries|to\s+everyone)\b|\bany\s+country\b|\binternational\s+(?:students\s+|applicants\s+)?(?:are\s+)?welcome\b|\bno\s+citizenship\s+requirement/i,
  usLocation:
    /\bmust\s+be\s+(?:located|based)\s+in\s+the\s+(?:U\.?S\.?|United States)\b|\bon-?site\b|\bin\s+person\b|\b(?:U\.?S\.?|United States)[- ]based\b|\benrolled\s+at\s+(?:a\s+)?(?:U\.?S\.?|United States)\s+(?:universit(?:y|ies)|institution|college)/i,
  remote: /\bremote\b|\bvirtual\b|\bonline\b/i,
  fundingRestricted:
    /\b(?:funding|stipend|award|fellowship|grant)\s+(?:is\s+)?(?:restricted\s+to|limited\s+to|available\s+only\s+to|only\s+for)\b[^.\n]*\b(?:citizens?|permanent\s+residents?)\b|\bNSF[- ]funded\b|\bfederal(?:ly)?\s+fund(?:ed|ing)\b/i,
  deadline:
    /\b(?:January|February|March|April|May|June|July|August|September|October|November|December)\s+\d{1,2},?\s+\d{4}\b|\b\d{4}-\d{2}-\d{2}\b|\b\d{1,2}\/\d{1,2}\/\d{4}\b/i,
  undergrad: /\bundergraduate\b|\bundergrads?\b|\bbachelor'?s?\b/i,
  graduate: /\bgraduate\s+students?\b|\bmaster'?s?\b|\bPhD\b|\bdoctoral\b/i,
  highSchool: /\bhigh\s+school\b/i,
  // Opportunity types
  internship: /\binternship\b|\bintern\b|\bco-?op\b/i,
  research: /\bresearch\b|\bREU\b|\blab\b/i,
  scholarship: /\bscholarship\b/i,
  fellowship: /\bfellowship\b/i,
  hackathon: /\bhackathon\b/i,
  competition: /\bcompetition\b|\bchallenge\b/i,
};

// ---------------------------------------------------------------------------
// Extraction
// ---------------------------------------------------------------------------

export function extractOpportunity(
  text: string,
  declaredType?: string
): ExtractedOpportunity {
  const evidence: EvidenceItem[] = [];
  const ambiguousPhrases: string[] = [];

  function record(field: string, value: string, d: Detection, confidence: number) {
    evidence.push({ field, value, source_text: d.source, confidence });
  }

  // --- Opportunity type ---
  let opportunityType = declaredType && declaredType !== "other" ? declaredType : "";
  if (!opportunityType) {
    const typeChecks: [string, RegExp][] = [
      ["internship", P.internship],
      ["scholarship", P.scholarship],
      ["fellowship", P.fellowship],
      ["hackathon", P.hackathon],
      ["research program", P.research],
      ["competition", P.competition],
    ];
    for (const [label, pattern] of typeChecks) {
      if (pattern.test(text)) {
        opportunityType = label;
        break;
      }
    }
    if (!opportunityType) opportunityType = "opportunity";
  }

  // --- Paid status ---
  let paidStatus: ExtractedOpportunity["paid_status"] = "unknown";
  const unpaidHit = detect(text, P.unpaid);
  const paidHit = detect(text, P.paid);
  if (unpaidHit) {
    paidStatus = "unpaid";
    record("paid_status", "unpaid", unpaidHit, 0.9);
  } else if (paidHit) {
    paidStatus = "paid";
    record("paid_status", "paid", paidHit, 0.92);
  }

  // --- Citizenship requirement ---
  let citizenship: string | null = null;
  const citizensOnlyHit = detect(text, P.citizensOnly);
  const prOnlyHit = detect(text, P.permanentResidentsOnly);
  const citizensOrPRHit = detect(text, P.citizensOrPR);
  if (citizensOnlyHit) {
    citizenship = "U.S. citizens only";
    record("citizenship_requirement", citizenship, citizensOnlyHit, 0.95);
  } else if (prOnlyHit) {
    citizenship = "Permanent residents only";
    record("citizenship_requirement", citizenship, prOnlyHit, 0.93);
  } else if (citizensOrPRHit) {
    citizenship = "U.S. citizens or permanent residents";
    record("citizenship_requirement", citizenship, citizensOrPRHit, 0.92);
  } else {
    citizenship = "not_explicit";
  }

  // --- Work authorization language ---
  let workAuthLanguage: string | null = null;
  const workAuthHit = detect(text, P.workAuth);
  const cptOptHit = detect(text, P.cptOpt);
  const ssnHit = detect(text, P.ssn);
  if (workAuthHit) {
    workAuthLanguage = workAuthHit.match;
    record("work_authorization_language", workAuthHit.match, workAuthHit, 0.95);
    ambiguousPhrases.push(workAuthHit.match);
  } else if (cptOptHit) {
    workAuthLanguage = cptOptHit.match;
    record("work_authorization_language", cptOptHit.match, cptOptHit, 0.85);
  }
  if (ssnHit) {
    if (!workAuthLanguage) workAuthLanguage = "SSN required";
    record("work_authorization_language", "SSN required", ssnHit, 0.88);
  }

  // --- Location / remote ---
  let location: string | null = null;
  let remoteStatus: string | null = null;
  const worldwideHit = detect(text, P.worldwide);
  const usLocationHit = detect(text, P.usLocation);
  const remoteHit = detect(text, P.remote);
  if (worldwideHit) {
    remoteStatus = "open worldwide";
    record("remote_or_global_status", "open worldwide", worldwideHit, 0.9);
  }
  if (usLocationHit) {
    location = "United States";
    record("location_requirement", "United States", usLocationHit, 0.88);
  }
  if (remoteHit && !worldwideHit) {
    remoteStatus = location ? "remote (U.S. only)" : "remote";
    record("remote_or_global_status", remoteStatus, remoteHit, 0.75);
  }

  // --- Funding restriction ---
  let fundingRestriction: string | null = null;
  const fundingHit = detect(text, P.fundingRestricted);
  if (fundingHit) {
    fundingRestriction = fundingHit.match;
    record("funding_restriction", fundingHit.match, fundingHit, 0.85);
  }

  // --- Deadline / start date ---
  let deadline: string | null = null;
  const deadlineHit = detect(text, P.deadline);
  if (deadlineHit) {
    deadline = deadlineHit.match;
    record("deadline_or_start_date", deadlineHit.match, deadlineHit, 0.9);
  }

  // --- Student level ---
  let studentLevel: string | null = null;
  if (P.highSchool.test(text)) studentLevel = "high school";
  else if (P.graduate.test(text)) studentLevel = "graduate";
  else if (P.undergrad.test(text)) studentLevel = "undergraduate";

  // --- International eligibility inference ---
  let intlEligibility: ExtractedOpportunity["international_eligibility"];
  if (citizensOnlyHit || prOnlyHit || citizensOrPRHit) {
    intlEligibility = "likely_not_eligible";
  } else if (worldwideHit) {
    intlEligibility = "likely_eligible";
  } else if (workAuthLanguage || paidStatus === "paid" || location) {
    intlEligibility = "unclear";
    if (workAuthLanguage && !ambiguousPhrases.includes(workAuthLanguage)) {
      ambiguousPhrases.push(workAuthLanguage);
    }
  } else {
    intlEligibility = "unknown";
  }

  return {
    opportunity_type: opportunityType,
    paid_status: paidStatus,
    work_authorization_language: workAuthLanguage,
    citizenship_requirement: citizenship,
    international_eligibility: intlEligibility,
    location_requirement: location,
    deadline_or_start_date: deadline,
    funding_restriction: fundingRestriction,
    student_level_requirement: studentLevel,
    remote_or_global_status: remoteStatus,
    required_materials: [],
    ambiguous_phrases: ambiguousPhrases,
    evidence,
  };
}

File: backend/services/extraction/phrase_rules.py
"""
Deterministic pre-LLM extraction layer.

PHRASE_RULES maps each ExtractedOpportunity field name to a list of pattern
descriptors. Each descriptor is matched against raw opportunity text before any
LLM call, producing zero-or-more EvidenceItem candidates with a normalized
label and a calibrated confidence score. The LLM pass later reconciles or
overrides these signals for ambiguous fields.
"""

import re
from typing import TypedDict


class RuleEntry(TypedDict):
    pattern: re.Pattern
    label: str
    confidence: float


def _r(pattern: str, label: str, confidence: float) -> RuleEntry:
    return {
        "pattern": re.compile(pattern, re.IGNORECASE),
        "label": label,
        "confidence": confidence,
    }


PHRASE_RULES: dict[str, list[RuleEntry]] = {
    # ------------------------------------------------------------------
    "paid_status": [
        _r(r"paid internship",              "paid",   0.95),
        _r(r"paid position",                "paid",   0.95),
        _r(r"paid summer",                  "paid",   0.95),
        _r(r"\bstipend\b",                  "paid",   0.95),
        _r(r"\bsalary\b",                   "paid",   0.95),
        _r(r"\bcompensation\b",             "paid",   0.95),
        _r(r"\$[\d,]+",                     "paid",   0.95),
        _r(r"fellows receive",              "paid",   0.95),
        _r(r"\bhourly\b",                   "paid",   0.95),
        _r(r"\bunpaid\b",                   "unpaid", 0.95),
        _r(r"\bvolunteer\b",                "unpaid", 0.95),
        _r(r"no compensation",              "unpaid", 0.95),
        _r(r"no payment",                   "unpaid", 0.95),
        _r(r"no employment or payment",     "unpaid", 0.95),
        # Negated lists like "No employment, compensation, or work authorization"
        _r(r"no employment[^.\n]{0,60}compensation", "unpaid", 0.96),
    ],

    # ------------------------------------------------------------------
    "work_authorization_language": [
        _r(r"eligible to work in the united states", "eligible to work in the united states", 0.97),
        _r(r"eligible to work in the u\.s",          "eligible to work in the u.s.",          0.97),
        _r(r"legally authorized to work",            "legally authorized to work",             0.97),
        _r(r"work authorization required",           "work authorization required",            0.97),
        _r(r"employment authorization",              "employment authorization",               0.97),
        _r(r"\bCPT\b",                               "CPT",                                   0.97),
        _r(r"\bOPT\b",                               "OPT",                                   0.97),
        _r(r"no work authorization required",        "no work authorization required",         0.97),
    ],

    # ------------------------------------------------------------------
    "citizenship_requirement": [
        _r(r"u\.s\. citizens only",                      "U.S. citizens only",                      0.97),
        _r(r"u\.s\. citizen",                            "U.S. citizen",                            0.97),
        _r(r"u\.s\. citizens or permanent residents",    "U.S. citizens or permanent residents",    0.97),
        _r(r"\bpermanent residents\b",                   "permanent residents",                     0.97),
        _r(r"\bgreen card holders\b",                    "green card holders",                      0.97),
        _r(r"international students are not eligible",   "international students are not eligible", 0.97),
        _r(r"non-u\.s\. citizens are not eligible",      "non-U.S. citizens are not eligible",      0.97),
        # Phrasings like "International students and students on temporary visas are not eligible"
        _r(r"international students[^.\n]{0,80}not eligible", "international students are not eligible", 0.97),
        _r(r"temporary visas?[^.\n]{0,60}not eligible",       "international students are not eligible", 0.97),
    ],

    # ------------------------------------------------------------------
    "international_eligibility": [
        _r(r"open worldwide",                        "likely_eligible", 0.95),
        _r(r"students worldwide",                    "likely_eligible", 0.95),
        _r(r"all countries",                         "likely_eligible", 0.95),
        _r(r"\bglobal\b",                            "likely_eligible", 0.95),
        _r(r"international students welcome",        "likely_eligible", 0.95),
        _r(r"international students are eligible",   "likely_eligible", 0.95),
        _r(r"\bworldwide\b",                         "likely_eligible", 0.95),
    ],

    # ------------------------------------------------------------------
    "location_requirement": [
        _r(r"\bremote\b",                       "remote",             0.90),
        _r(r"\bonline\b",                       "online",             0.90),
        _r(r"\bhybrid\b",                       "hybrid",             0.90),
        _r(r"\bin-person\b",                    "in-person",          0.90),
        _r(r"\bunited states\b",                "United States",      0.90),
        _r(r"\bu\.s\.\b",                       "U.S.",               0.90),
        _r(r"must be located in the u\.s",      "must be located in the U.S.", 0.90),
        _r(r"open worldwide",                   "worldwide",          0.90),
    ],

    # ------------------------------------------------------------------
    "funding_restriction": [
        _r(r"funding restricted to u\.s\. citizens",  "funding restricted to U.S. citizens", 0.92),
        _r(r"\bfederal funding\b",                    "federal funding",                     0.92),
        _r(r"\bNSF REU\b",                            "NSF REU",                             0.92),
        _r(r"funding may be available",               "funding may be available",            0.92),
        _r(r"citizenship-based funding",              "citizenship-based funding",           0.92),
        # "Funding is restricted to ..." / "funding restricted by ..."
        _r(r"funding (?:is |may be )?restricted",     "funding restricted by citizenship or residency", 0.92),
    ],

    # ------------------------------------------------------------------
    "student_level_requirement": [
        _r(r"\bhigh school\b",              "high school",                0.88),
        _r(r"\bundergraduate\b",            "undergraduate",              0.88),
        _r(r"\bgraduate\b",                 "graduate",                   0.88),
        _r(r"\bfreshman\b",                 "freshman",                   0.88),
        _r(r"\bsophomore\b",                "sophomore",                  0.88),
        _r(r"\bjunior\b",                   "junior",                     0.88),
        _r(r"\bsenior\b",                   "senior",                     0.88),
        _r(r"\bcollege students\b",         "college students",           0.88),
        _r(r"accredited u\.s\. institutions","accredited U.S. institutions",0.88),
        _r(r"u\.s\. universities",          "U.S. universities",          0.88),
    ],

    # ------------------------------------------------------------------
    "deadline_or_start_date": [
        _r(
            r"\b(january|february|march|april|may|june|july|august|september|october|november|december)\s+\d{1,2},?\s+\d{4}\b",
            "date_mention",
            0.90,
        ),
        _r(r"\b\d{1,2}/\d{1,2}/\d{4}\b",  "date_mention",       0.90),
        _r(r"application deadline",         "application deadline",0.90),
        _r(r"applications due",             "applications due",   0.90),
        _r(r"\bstart date\b",              "start date",         0.90),
        _r(r"\bsummer 2026\b",             "summer 2026",        0.90),
    ],

    # ------------------------------------------------------------------
    "required_materials": [
        _r(r"\bresume\b",                  "resume",               0.93),
        _r(r"\bCV\b",                      "CV",                   0.93),
        _r(r"\btranscript\b",              "transcript",           0.93),
        _r(r"\bcover letter\b",            "cover letter",         0.93),
        _r(r"\bgithub repository\b",       "github repository",    0.93),
        _r(r"\bportfolio\b",               "portfolio",            0.93),
        _r(r"\brecommendation letter\b",   "recommendation letter",0.93),
        _r(r"\bessay\b",                   "essay",                0.93),
        _r(r"\bproject report\b",          "project report",       0.93),
        _r(r"\bpresentation\b",            "presentation",         0.93),
        _r(r"\bvideo demo\b",              "video demo",           0.93),
        _r(r"\bstatement of interest\b",   "statement of interest",0.93),
    ],

    # ------------------------------------------------------------------
    "ambiguous_phrases": [
        _r(r"funding may be available",     "funding may be available",    0.70),
        _r(r"must be eligible",             "must be eligible",            0.70),
        _r(r"authorization required",       "authorization required",      0.70),
        _r(r"u\.s\. institutions",          "u.s. institutions",           0.70),
        _r(r"remote but must be located",   "remote but must be located",  0.70),
        _r(r"work with faculty mentors",    "work with faculty mentors",   0.70),
        _r(r"\blegal authorization\b",      "legal authorization",         0.70),
        _r(r"sponsorship not provided",     "sponsorship not provided",    0.70),
    ],

    # ------------------------------------------------------------------
    "opportunity_type": [
        _r(r"\bhackathon\b",    "hackathon",   0.95),
        _r(r"\binternship\b",   "internship",  0.95),
        _r(r"\bfellowship\b",   "fellowship",  0.95),
        _r(r"\bresearch\b",     "research",    0.95),
        _r(r"\bcompetition\b",  "competition", 0.95),
        _r(r"\bscholarship\b",  "scholarship", 0.95),
    ],
}
